src/packet/ipv6.rs

```diff
@@ -32,6 +32,86 @@ pub struct Ipv6 {
     payload: Vec<u8>,
 }
 
+/// Represents an IPv6 Packet pseudo header
+#[packet] pub struct Ipv6PseudoHeader {
+    #[construct_with(u16, u16, u16, u16, u16, u16, u16, u16)]
+    source: Ipv6Addr,
+    // RFC 2460:
+    // If the IPv6 packet contains a Routing header, the Destination Address used in the
+    // pseudo-header is that of the final destination. At the originating node, that address will
+    // be in the last element of the Routing header; at the recipient(s), that address will be in
+    // the Destination Address field of the IPv6 header.
+    #[construct_with(u16, u16, u16, u16, u16, u16, u16, u16)]
+    destination: Ipv6Addr,
+    // RFC 2460: 
+    // The Upper-Layer Packet Length in the pseudo-header is the length of the
+    // upper-layer header and data (e.g., TCP header plus TCP data). Some upper-layer protocols
+    // carry their own length information (e.g., the Length field in the UDP header); for such
+    // protocols, that is the length used in the pseudo- header. Other protocols (such as TCP) do
+    // not carry their own length information, in which case the length used in the pseudo-header
+    // is the Payload Length from the IPv6 header, minus the length of any extension headers
+    // present between the IPv6 header and the upper-layer header.
+    inner_packet_length: u32be,
+    zeros: u24be,
+    // RFC 2460:
+    // The Next Header value in the pseudo-header identifies the upper-layer protocol (e.g., 6 for
+    // TCP, or 17 for UDP). It will differ from the Next Header value in the IPv6 header if there
+    // are extension headers between the IPv6 header and the upper-layer header.
+    #[construct_with(u8)]
+    next_level_protocol: IpNextHeaderProtocol,
+    #[payload]
+    payload: Vec<u8>,
+}
+
+pub struct Ipv6HeaderIter<'p, 'h> {
+    packet: &'p Ipv6Packet<'p>,
+    current_header: Ipv6Packet<'h>,
+}
+
+impl<'p, 'h> Iterator for Ipv6HeaderIter<'p, 'h> {
+    type Item = &'h Ipv6Packet<'h>;
+    fn next(&mut self) -> Option<&'h Ipv6Packet<'h>> {
+        match self.current_header.get_next_header() {
+            IpNextHeaderProtocols::Hopopt |
+            IpNextHeaderProtocols::Ipv6 |
+            IpNextHeaderProtocols::Ipv6Route | 
+            IpNextHeaderProtocols::Ipv6Frag |
+            IpNextHeaderProtocols::Ipv6Icmp |
+            IpNextHeaderProtocols::Ipv6Opts => {
+                self.current_header = Ipv6Packet::new(self.current_header.payload()).unwrap(); // is it ok to unwrap here?
```

Comment:
You probably need to `use packet::Packet`.

Calling `.unwrap()` is safe, providing `self.current_header.payload().len() >= Ipv6Packet::minimum_packet_size()`.

Relatedly, though not particularly importantly, it's a shame you have to write out this code, it seems like something that pnet_macros should be able to do for you. There's currently functionality for handling optional fields (ie. IPv4 options), but obviously each potential next header field for IPv6 has a different packet type. It might be worth investigating at some point to see if it can be automated or not.